### DISCUSSION TITLE
Intel AMT activation on remote devices in hosted server [ internet ]

### DISCUSSION THREAD

**Discussion by amoljagdalepucsd** [Ylianst/MeshCentral#5899] [Q&A]
Hi community,
I am trying to activate AMT for a remote session. So, I ran MeshCentral [version 1.0.85] locally and created groups, both agent-specific and AMT-specific, with CCM and ACM policy with CIRA, and it worked fine. 
I put credentials for ACM mode and was able to access and see all the things on the localhost server and in the same network. However, I have hosted MeshCentral on https://abc.com/. ( I am using Kubernetes cluster where I have configure Ingress )
I created agent-specific groups and onboarded devices under those groups, trying to activate AMT using ACM using MEBx credentials, but it's not getting activated. 
I tried with CCM as well, but I'm not making progress. 
I want to activate Intel AMT on devices which is onboarded in internet-hosted server [https://abc.com/]. 
 I checked the logs and am getting an error at the below line in amtscanner.js:
**obj.checkTcpPresence(host, (doc.intelamt.tls == 1) ? 16993 : 16992, scaninfo, function (tag, result, version) { }**
Here **doc.itemamt** is undefined, so it's throwing the below error:
**SQLERR5 TypeError: Cannot read properties of undefined (reading 'tls')**
I'm not sure if this is the root cause for not activating AMT on remote devices on the hosted server. 

**Adding some details:** 
 config.JSON file:
```
{
    "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
    "settings": {
        "cert": "https://abc.com",
        "_WANonly": false,
        "_LANonly": false,
        "_sessionKey": "MyReallySecretPassword1",
        "sessionSameSite": "lax",
        "port": 4430,
        "aliasPort": 443,
        "_redirPort": 800,
        "_redirAliasPort": 80,
        "agentPort": 2121,
        "agentAliasPort": 443,
        "_agentAliasDNS": "https://abc.com",
        "relayPort": 453,
        "_relayAliasPort": 453,
        "_relayDNS": "https://abc.com",
        "agentPortTls": false,
        "agentPing": 10,
        "browserPing": 10,
        "agentIdleTimeout": 18000,
        "tlsOffload": true,
        "SelfUpdate": false,
        "allowFraming": true,
        "allowLoginToken": true,
        "webRTC": true,
        "multitenant": false,
        "amtscanner": true,
        "amtManager": true,
        "plugins": {
            "enabled": true
        },
        "postgres": {
            "user": "admin",
            "password": "aadfxmtydfd",
            "port": "5432",
            "host": "1.2.3.4.5",
            "database": "meshdb",
            "ssl": false
        },
        "dbExpire": {
            "events": 0,
            "powerevents": 0,
            "statsevents": 0
        },
        "agentCoreDump": true,
        "agentLogDump": true,
        "minify": true,
        "clipboardGet": true,
        "clipboardSet": true,
        "compression": true,
        "wsCompression": true,
        "agentWsCompression": true,
        "noAgentUpdate": 1,
        "temporaryAgentUpdate": false,
        "agentsInRam": true,
        "desktopMultiplex": true
    },
    "domains": {
        "": {
            "title": "Meshcentral",
            "title2": "Mesh service",
            "minify": "true",
            "NewAccounts": false,
            "_userNameIsEmail": true,
            "certUrl": "https://abc.com",
            "tlsOffload": "true",
            "notifyimage": "abc.png",
            "meshMessengerTitle": "MeshMessenger",
            "nightmode": 1,
            "userSessionIdleTimeout": 90,
            "amtmanager": {
                "tlsconnections": false,
                "adminaccounts": [
                    {
                        "user": "admin",
                        "pass": "Remote@12345"
                    }
                ]
            },
            "_amtScanOptions": []
        }
    }
}
```

I entered some commands in the agent console as well. Below are the details:
**Command: amt 
response :** {
    core-ver: 1
    OsHostname: "ILD-PUN-LAP-084"
    Flags: 2
    MeiVersion: "100.0.0.5124"
    Versions: {
        Flash: "11.0.18"
        Netstack: "11.0.18"
        AMTApps: "11.0.18"
        AMT: "11.0.18"
        Sku: "16392"
        VendorID: "8086"
        Build Number: "1002"
        Recovery Version: "11.0.18"
        Recovery Build Num: "1002"
        Legacy Mode: "False"
    }
    UUID: "4c4c4544-0048-4810-8051-c3c04f5a4632"
    ProvisioningMode: 1
    ProvisioningState: 2
    net0: {
        enabled: 1
        dhcpEnabled: 1
        dhcpMode: "PASSIVE"
        mac: "D4:81:D7:A4:94:5A"
        address: "0.0.0.0"
    }
    net1: {
        enabled: 1
        dhcpEnabled: 1
        dhcpMode: "PASSIVE"
        mac: "28:16:AD:8A:1E:1B"
        address: "0.0.0.0"
    }
}

**Command: amtconfig** 
response :
12:52:55, LMS tunnel start.
12:52:56, LMS tunnel closed.

**Command: amtevents** 
response : 
12:21:15, LMS tunnel start.
12:21:16, LMS tunnel closed.
12:24:32, LMS tunnel start.
12:24:33, LMS tunnel closed.
12:27:58, LMS tunnel start.
12:27:59, LMS tunnel closed.
12:52:55, LMS tunnel start.
12:52:56, LMS tunnel closed.

Please suggest if I missed anything.

**Comment by si458** [comment]
You are running an old version (the was a few intel amt bug fixes in a recent update)
1. update the latest version and try again.
2. use a newer nodejs like 18 or above (if u aren't already)

**Comment by amoljagdalepucsd** [comment]
Ok But its working locally with this version @si458

**Reply by si458** [reply]
Ohh just read ur config (sorry just woke up)

You actually have a BUG!?

You have set `tlsconnections: false` in your config.json so oversally tls will be undefined!?

You need the tls enabled for Internet amt

**Reply by amoljagdalepucsd** [reply]
Okay got it. let me retry it and update here

**Reply by amoljagdalepucsd** [reply]
Enabled it but still getting same error. 
while activating AMT do we need to change any other keys in config.JSON. for example tlsOffload ?
 @si458

**Reply by si458** [reply]
can you run `info` from the 'server console` and let me have the output?

**Reply by amoljagdalepucsd** [reply]
command : info
> info
Current Core: Jan 18 2024, 1951026568
Agent Time: 2024-03-05 01:36:27.234-08:00.
User Rights: 0xffffffff.
Platform: win32.
Capabilities: 15.
Server URL: wss://abc.com:443/agent.ashx.
Built-in LMS: Disabled.
OS: Microsoft Windows 10 Enterprise - 22H2/19045.
Modules: amt-apfclient, amt-lme, amt-manage, amt-mei, monitor-border, smbios, sysinfo, util-agentlog, wifi-scanner-windows, wifi-scanner, win-console, win-info, win-securitycenter, win-terminal, win-virtual-terminal, win-volumes.
Server Connection: true, State: 1.

**Reply by si458** [reply]
thats a remote device info not the server info, use the 'My Server' tab and its console for the `info` output
also you have rebuilt the meshagent yourself? (only cuz the current core, is very new haha)

**Reply by amoljagdalepucsd** [reply]
Ok. {
    "meshVersion": "v1.0.85",
    "nodeVersion": "v20.11.1",
    "runMode": "Hybrid (LAN + WAN) mode",
    "productionMode": true,
    "database": "PostgreSQL",
    "plugins": [],
    "platform": "linux",
    "arch": "x64",
    "pid": 47,
    "uptime": 1695.766394147,
    "cpuUsage": {
        "user": 47980456,
        "system": 4699920
    },
    "memoryUsage": {
        "rss": 339062784,
        "heapTotal": 88629248,
        "heapUsed": 72520944,
        "external": 218059860,
        "arrayBuffers": 215747661
    },
    "warnings": [],
    "allDevGroupManagers": []
}

**Reply by si458** [reply]
i would try updating ur mesh as its still showing 1.0.85, we are on the 1.1.21 and i remember around 1.1.19 we did AMT fixes which i discovered with newer nodejs versions...

**Reply by amoljagdalepucsd** [reply]
Ok. but still not sure my config.JSON is correct and enabled all required things

**Reply by si458** [reply]
Did u try updating to 1.1.21?

**Reply by amoljagdalepucsd** [reply]
Hi @si458 
I investigated the code further regarding the 'doc.intelamt.tls undefined' issue which i mentioned in above description and determined that it is not the root cause of the failure to activate AMT. 
obj.checkTcpPresence(host, (doc.intelamt.tls == 1) ? 16993 : 16992, scaninfo, function (tag, result, version) { }
The error occurs because 'doc.intelamt' is undefined, leading to the following error:
**SQLERR5 TypeError: Cannot read properties of undefined (reading 'tls')**
I discovered that 'doc' refers to a node object, and in the 'amtscanner.js' file, it scans local AMT nodes from the hosted server by fetching records from the database. Then, it performs TCP detection on a port based on the 'doc.intelamt.tls' condition.
The condition is:
obj.checkTcpPresence(host, (doc.intelamt.tls == 1) ? 16993 : 16992,

Therefore, it attempts to check the 'doc.intelamt.tls' property, which results in 'undefined.tls' for some nodes where 'doc.intelamt' itself is undefined. Consequently, the error is thrown.
The 'intelamt' property is not available for some inactive devices or devices with outdated records in the database. I have handled this issue, and it no longer throws an error.
I am attempting to activate Intel AMT using MEbx credentials on remote devices onboarded on a hosted server over the internet. I'm using a Kubernetes cluster where I've configured Ingress and Certificate 1.2 to route requests to MeshCentral. Do we need any other certificate for AMT? Additionally, the remote machines are in different networks.
Since we are using MeshCentral with a hosted server and some machines are onboarded for our use, if we update and it breaks, it will cause issues. Therefore, we plan to update it next week. However, today I reviewed Intel AMT fixes from the 1.0.85 to 1.1.21 release and incorporated all changed files into version 1.0.85, but the AMT is still not getting activated.

as I'm relatively new to this, if there are any aspects I've overlooked due to my limited understanding, please provide guidance.
Thanks

**Reply by si458** [reply]
You can't just apply patches like that...
You must actually update the npm package to 1.1.21!

**Comment by thermionic** [comment]
I'm intrigued as to how version 1.085 was deployed, is this is a 3rd party container image?
